ocaml/forkexecd/lib/forkhelpers.ml

```diff
@@ -72,14 +74,50 @@ let waitpid (sock, pid) =
       in
       failwith msg
 
-let waitpid_nohang ((sock, _) as x) =
+(* [waitpid_nohang] reports the status of a socket to a process. The
+   intention is to make this non-blocking. If the process is finished,
+   the socket is closed by default and not otherwise. The [close] option
+   can be used to never close the socket such that this is left to the
+   caller, which is a cleaner protocol. *)
+let waitpid_nohang ?(close = true) (sock, pid) =
+  D.debug "%s pid=%d" __FUNCTION__ pid ;
+  let fail fmt = Printf.kprintf failwith fmt in
   Unix.set_nonblock sock ;
-  let r =
-    try waitpid x
-    with Unix.(Unix_error ((EAGAIN | EWOULDBLOCK), _, _)) ->
-      (0, Unix.WEXITED 0)
-  in
-  Unix.clear_nonblock sock ; r
+  match Fecomms.read_raw_rpc sock with
+  | Ok Fe.(Finished (WEXITED n)) ->
+      D.debug "%s pid=%d WEXITED" __FUNCTION__ pid ;
+      Unix.clear_nonblock sock ;
+      if close then Unix.close sock ;
+      (pid, Unix.WEXITED n)
+  | Ok Fe.(Finished (WSIGNALED n)) ->
+      D.debug "%s pid=%d WSIGNALED" __FUNCTION__ pid ;
+      Unix.clear_nonblock sock ;
+      if close then Unix.close sock ;
+      (pid, Unix.WSIGNALED n)
+  | Ok Fe.(Finished (WSTOPPED n)) ->
+      D.debug "%s pid=%d WSTOPPED" __FUNCTION__ pid ;
+      Unix.clear_nonblock sock ;
+      if close then Unix.close sock ;
+      (pid, Unix.WSTOPPED n)
+  | Ok status ->
+      Unix.clear_nonblock sock ;
+      fail "%s: unexpected status received (%s)" __FUNCTION__
+        (Fe.ferpc_to_string status)
+  | Error msg ->
+      D.debug "%s pid=%d %s" __FUNCTION__ pid msg ;
+      Unix.clear_nonblock sock ;
+      fail "%s: error happened when trying to read the status. %s" __FUNCTION__
+        msg
+  (* it's a bit crazy that we have Result.t and exceptions from
+     read_raw_rpc *)
+  | exception Unix.(Unix_error ((EAGAIN | EWOULDBLOCK), _, _)) ->
+      D.debug "%s pid=%d EAGAIN EWOULDBLOCK" __FUNCTION__ pid ;
+      Unix.clear_nonblock sock ;
+      (0, Unix.WEXITED 0) (* is this a convention with the caller? *)
+  | exception exn ->
+      D.debug "%s pid=%d %s" __FUNCTION__ pid (Printexc.to_string exn) ;
+      fail "%s: error happened when trying to read the status. %s" __FUNCTION__
```

Comment:
The `Unix.clear_nonblock sock ;` is missing here, shouldn't it be restored like the other cases?

Reply:
We could try as a best effort. At this point it's not clear that the socket is still open and we are in an unknown state.